nsqd/lookup.go

```diff
@@ -118,35 +133,68 @@ func (n *NSQD) lookupLoop() {
 			}
 		case val := <-n.notifyChan:
 			var cmd *nsq.Command
+			var err error
 			var branch string
-
-			switch val.(type) {
-			case *Channel:
-				// notify all nsqlookupds that a new channel exists, or that it's removed
-				branch = "channel"
-				channel := val.(*Channel)
-				if channel.Exiting() == true {
-					cmd = nsq.UnRegister(channel.topicName, channel.name)
-				} else {
-					cmd = nsq.Register(channel.topicName, channel.name)
-				}
-			case *Topic:
-				// notify all nsqlookupds that a new topic exists, or that it's removed
-				branch = "topic"
-				topic := val.(*Topic)
-				if topic.Exiting() == true {
-					cmd = nsq.UnRegister(topic.name, "")
-				} else {
-					cmd = nsq.Register(topic.name, "")
+			notifyCtx, ok := val.(notifyContext)
+			if ok {
+				switch notifyCtx.notifyType {
+				case NotifyTypeRegistration:
+					switch notifyCtx.v.(type) {
+					case *Channel:
+						// notify all nsqlookupds that a new channel exists
+						branch = "channel"
+						channel := notifyCtx.v.(*Channel)
+						cmd = nsq.Register(channel.topicName, channel.name)
+					case *Topic:
+						// notify all nsqlookupds that a new topic exists
+						branch = "topic"
+						topic := notifyCtx.v.(*Topic)
+						cmd = nsq.Register(topic.name, "")
+					}
+				case NotifyTypeUnRegistration:
+					switch notifyCtx.v.(type) {
+					case *Channel:
+						// notify all nsqlookupds that a new channel removed
+						branch = "channel"
+						channel := notifyCtx.v.(*Channel)
+						cmd = nsq.UnRegister(channel.topicName, channel.name)
+					case *Topic:
+						// notify all nsqlookupds that a new topic removed
+						branch = "topic"
+						topic := notifyCtx.v.(*Topic)
+						cmd = nsq.UnRegister(topic.name, "")
+					}
+				case NotifyTypeStateUpdate:
+					switch notifyCtx.v.(type) {
+					case *Channel:
+						// notify all nsqlookupds that channel state changed
+						branch = "channel"
+						channel := notifyCtx.v.(*Channel)
+						cmd, err = nsq.SyncState(channel.topicName, channel.name, map[string]interface{}{"paused": channel.IsPaused()})
+						if err != nil {
+							n.logf(LOG_ERROR, "NSQD: build cmd err: %s", err)
+						}
+					case *Topic:
+						// notify all nsqlookupds that topic state changed
+						branch = "topic"
+						topic := notifyCtx.v.(*Topic)
+						cmd, err = nsq.SyncState(topic.name, "", map[string]interface{}{"paused": topic.IsPaused()})
+						if err != nil {
+							n.logf(LOG_ERROR, "NSQD: build cmd err: %s", err)
+						}
+					}
+				default:
+					n.logf(LOG_ERROR, "NSQD: unknown notifyType: %d", notifyCtx.notifyType)
```

Comment:
Same here, this should be a `panic` because it's essentially a programming error and behavior is undefined.

Reply:
done.